### ISSUE TITLE
Ceph external cluster, repeatedly saving mon endpoints to config

### ISSUE THREAD

**Issue by eliaswimmer** [rook/rook#4717]
**Is this a bug report or feature request?**

* Bug Report

  
**Bug Report**

Rook rewrites configs for external ceph cluster repeatedly every few seconds.

**Expected behavior:**

Only rewrite configs when cluster changes.

**Deviation from expected behavior:**
```
2020-01-17 14:03:25.162995 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:03:25.163036 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:03:25.163052 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:03:25.163064 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:03:25.173956 I | op-mon: saved mon endpoints to config map map[data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}]]
2020-01-17 14:03:25.180472 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:03:25.180747 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:04:11.090181 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:04:11.090223 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:04:11.090238 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:04:11.090249 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:04:11.100412 I | op-mon: saved mon endpoints to config map map[maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789]
2020-01-17 14:04:11.106600 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:04:11.106910 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:04:56.987005 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:04:56.987059 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:04:56.987075 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:04:56.987087 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:04:56.997593 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}}]
2020-01-17 14:04:57.003829 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:04:57.004133 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:05:42.858947 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:05:42.859009 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:05:42.859028 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:05:42.859044 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:05:42.871217 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}}]
2020-01-17 14:05:42.877706 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:05:42.877999 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:06:28.772799 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:06:28.772840 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:06:28.772855 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:06:28.772868 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:06:28.784259 I | op-mon: saved mon endpoints to config map map[mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.7:6789","10.1.1.5:6789","10.1.1.6:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1]
2020-01-17 14:06:28.789822 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:06:28.790111 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
```

**How to reproduce it (minimal and precise):**

Use rook 1.2.1 with a external cluster with 3 mons.
```
apiVersion: v1
kind: Namespace
metadata:
  name: {{ rook_namespace }}
---
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-mon
  namespace: {{ rook_namespace }}
type: Opaque
data:
  cluster-name: {{ rook_clustername | b64encode }}
  fsid: {{ rook_external_fsid | b64encode }}
  admin-secret: {{ rook_external_admin_secret | b64encode }}
  mon-secret: {{ rook_external_mon_secret | b64encode }}
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: rook-ceph-mon-endpoints
  namespace: {{ rook_namespace }}
data:
  data: "{{ rook_external_ceph_mon_data }}"
  mapping: "{}"
  maxMonId: "2"
---
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: {{ rook_clustername }}
  namespace: {{ rook_namespace }}
spec:
  external:
    enable: true
  dataDirHostPath: /var/lib/rook
  # providing an image is optional, do this if you want to create other CRs (rgw, mds, nfs)
  cephVersion:
    image: ceph/ceph:v14.2.6 # MUST match external cluster version
```
**Environment**:
* OS: CentOS 7
* Kernel : 3.10.0-1062.9.1.el7.x86_64
* Cloud provider or hardware configuration: Bare Metal
* Rook version: v1.2.1
* Storage backend version: ceph 14.2.6
* Kubernetes version: v1.17.0
* Kubernetes cluster type: Bare Metal
* Storage backend status: HEALTH_OK

**Comment by leseb** [comment]
@rajatsingh25aug PTAL this is due to https://github.com/rook/rook/pull/4150.

**Comment by rajatsingh25aug** [comment]
It's happening due to this, this was written in this way only.
https://github.com/rajatsingh25aug/rook/blob/65fb5684f28854def6ac719ef0e3e079efa76a87/pkg/operator/ceph/cluster/mon/health.go#L330
I don't think

**Comment by rajatsingh25aug** [comment]
cc @eliaswimmer can you please check the link that I provided?.
Thanks

**Comment by eliaswimmer** [comment]
@rajatsingh25aug Yes that's the function rewriting the config over and over.
I had the time to learn some go over the weekend, I will try to fix this.
I think the indented functionality was to only write the config when some mons join or leave.

**Comment by rajatsingh25aug** [comment]
Yes that's absolutely correct. But l think that modifying this might break
the code from several places.
But feel free to raise a PR and I'll review it.
Thanks

**Comment by leseb** [comment]
This issue is important, @rajatsingh25aug please keep track of it. We need to backport this fix in 1.2.

@eliaswimmer do you commit to fixing this issue? Thanks.

**Comment by eliaswimmer** [comment]
@leseb @rajatsingh25aug Please review my PR
